src/full_node/block_header_validation.py

```diff
@@ -1,260 +1,660 @@
-        return Err.INVALID_NUM_ITERATIONS
-
-    # 2. the PoT must be valid, on a discriminant of size 1024, and the challenge_hash
-    if not pre_validated:
-        if not block.proof_of_time.is_valid(constants.DISCRIMINANT_SIZE_BITS):
-            return Err.INVALID_POT
-    # 3. If not genesis, the challenge_hash in the proof of time must match the challenge on the previous block
-    if not genesis:
-        assert prev_header_block is not None
-        prev_challenge: Optional[Challenge] = prev_header_block.challenge
-        assert prev_challenge is not None
-
-        if block.proof_of_time.challenge_hash != prev_challenge.get_hash():
-            return Err.INVALID_POT_CHALLENGE
-    else:
-        # 9. If genesis, the challenge hash in the proof of time must be the same as in the proof of space
-        assert block.proof_of_time is not None
-        challenge_hash = block.proof_of_time.challenge_hash
-
-        if challenge_hash != block.proof_of_space.challenge_hash:
-            return Err.INVALID_POSPACE_CHALLENGE
-
-    return None
-
-
-def pre_validate_finished_block_header(constants: ConsensusConstants, data: bytes):
+    header_block: HeaderBlock,
+    check_filter: bool,
+) -> Tuple[Optional[uint64], Optional[Err]]:
     """
-    Validates all parts of block that don't need to be serially checked
+    Fully validates the header of a sub-block. A header block is the same as a full block, but
+    without transactions and transaction info. Returns (required_iters, error).
     """
-    block = FullBlock.from_bytes(data)
-
-    if not block.proof_of_time:
-        return False, None
 
-    # 4. Check PoT
-    if not block.proof_of_time.is_valid(constants.DISCRIMINANT_SIZE_BITS):
-        return False, None
+    unfinished_header_block = UnfinishedHeaderBlock(
+        header_block.finished_slots,
+        header_block.reward_chain_sub_block.get_unfinished(),
+        header_block.challenge_chain_icp_proof,
+        header_block.reward_chain_icp_proof,
+        header_block.foliage_sub_block,
+        header_block.foliage_block,
+        header_block.transactions_filter,
+    )
 
-    # 9. Check harvester signature of header data is valid based on harvester key
-    validates = blspy.AugSchemeMPL.verify(
-        block.proof_of_space.plot_public_key,
-        block.header.data.get_hash(),
-        block.header.plot_signature,
+    required_iters, validate_unfinished_result = await validate_unfinished_header_block(
+        constants, sub_blocks, height_to_hash, unfinished_header_block, check_filter
     )
-    if not validates:
-        return False, None
 
-    # 10. Check proof of space based on challenge
-    pos_quality_string = block.proof_of_space.verify_and_get_quality_string(
-        constants.NUMBER_ZERO_BITS_CHALLENGE_SIG
+    if validate_unfinished_result is not None:
+        return None, validate_unfinished_result
+    if header_block.height == 0:
+        prev_sb: Optional[SubBlockRecord] = None
+    else:
```

Comment:
also i think we can do something like:

`  if header_block.height == 0:
        prev_sb: Optional[SubBlockRecord] = None
        ips = uint64(constants.IPS_STARTING)
    else:
        prev_sb: Optional[SubBlockRecord] = sub_blocks[header_block.prev_header_hash]
        ips: uint64 = get_next_ips(constants, sub_blocks, height_to_hash, header_block.prev_header_hash, new_slot)
              ip_iters: uint64 = calculate_ip_iters(constants, ips, required_iters)`

Reply:
i can change it myself, dont want to create conflicts by working on the same file, lmk

Reply:
Feel free to make the changes yourself

Reply:
There are a lot of simplifications like that, that we can make